### PR TITLE
[In-progress] MSYS-618 - Performance tunning of Install-Project

### PR DESCRIPTION
**Pull request by piyushawasthi**
Here is  Bench-Mark after these changes.

**Download and install Chef-client With daemon option is service**

VERBOSE: Total Minutes -> 1
VERBOSE: Total Second -> 12

**Download and install Chef-client With daemon option is service on FIPS**

VERBOSE: Total Minutes -> 1
VERBOSE: Total Second -> 19

### CODE REVIEW

**Comment by wrightp** [comment]
@piyushawasthi I temporarily switched the merge from master to a custom branch so I could initiate the acceptance tests. I will move it back once they pass. Also, not sure why the travis tests are failing. Seems unrelated to your changes.

**Comment by wrightp** [comment]
The test failures appear to be related to transitive deps. Looking into this. I ran the acceptance tests locally and they passed.

**Comment by wrightp** [comment]
@mwrock Could you please take a moment to look at this PR? Will this have any effect on nano systems?

File: lib/mixlib/install/generator/powershell/scripts/helpers.ps1.erb [hunk 1]
```diff
@@ -110,20 +118,25 @@ function Get-WebContentOnCore {
 function Test-ProjectPackage {
   [cmdletbinding()]
   param ($Path, $Algorithm = 'SHA256', $Hash)
+  if (!$env:Valid_ProjectPackage){
+    Write-Verbose "Testing the $Algorithm hash for $path."
+    $ActualHash = if (Is-FIPS) {
+      (Custom-GetFileHash -Algorithm $Algorithm -Path $Path).Hash.ToLower()
+    } else {
+      (Get-FileHash -Algorithm $Algorithm -Path $Path).Hash.ToLower()
```
Comment (mwrock) [inline lib/mixlib/install/generator/powershell/scripts/helpers.ps1.erb]: does the `get-filehash` function still exist?

**Comment by mwrock** [comment]
Have you done perf comparisons looking at time taken with this change and without?

Caching in the environment will only work if you are making multiple calls inside of the same powershell invokation.

File: lib/mixlib/install/generator/powershell/scripts/helpers.ps1.erb (hunk 1 above)
Comment (piyushawasthi) [inline lib/mixlib/install/generator/powershell/scripts/helpers.ps1.erb]: No, Here `Get-FileHash` is powershell utility.  I have renamed  `Get-FileHash` into `Custom-GetFileHash`. I just make less use of that function due to N+ loop between `disposable` and `use` function.

**Comment by piyushawasthi** [comment]
@mwrock - I have caching the environment because there are 2 calls for the same. Also Without these changes `Install-Project -project chef -channel stable -daemon service -Verbose` take time upto 2 minutes (Avg between 1.39 min to 1.53 min) now it will reach around 1 min (between 0.58 to 1.1). 

**Sample snapshot after changes:**
final_reduce

File: lib/mixlib/install/generator/powershell/scripts/helpers.ps1.erb (hunk 1 above)
Comment (mwrock) [inline lib/mixlib/install/generator/powershell/scripts/helpers.ps1.erb]: ahh. right. be aware that cmdlet was introduced in ps v4 and will therefore fail on OS versions prior to 2012R2 unless their ps has been updated.

**Comment by mwrock** [comment]
thanks for the explanation above. That makes sense.

**Comment by piyushawasthi** [comment]
Thanks @mwrock

**Comment by mwrock** [comment]
So just to reiterate the above comment about `Get-FileHash`. That will break on windows 2012 (not R2) or 2008/2008 R2.

**Comment by wrightp** [comment]
@mwrock  Oooh Snap. I need to revert this change in master. @piyushawasthi There will be work to do yes?

**Comment by piyushawasthi** [comment]
@wrightp: I have done with changes but I am verifying on different platform to ensure no issue on any platform. I'll let you know ASAP.

File: lib/mixlib/install/generator/powershell/scripts/helpers.ps1.erb (hunk 1 above)
Comment (piyushawasthi) [inline lib/mixlib/install/generator/powershell/scripts/helpers.ps1.erb]: make sense, Let me verify on ps v2 and ps v3.

**Pull request merged**